### PR TITLE
Preserve display plots on update, instead of trashing them all

### PR DESCRIPTION
**Pull request by chaosphere2112**
In working on the GUI refactor, I realized I was dedicating a ton of effort to managing the refresh of display plots. I poked my head into the guts of the update/clear cycle and the display plot generation inside of `canvas.__plot` and it wasn't too hard to tweak it to actually keep the same object, so that's precisely what I've done. No rush on merging this, certainly not necessary for the release; I want to make sure it's not doing anything too wacky before it goes anywhere. Just thought I'd submit this so it's available for scrutiny.

### CODE REVIEW

**Comment by doutriaux1** [comment]
@chaosphere2112 do we want to merge this?

**Comment by aashish24** [comment]
If someone is clearing the canvas, then what we are preserving the display?

**Comment by chaosphere2112** [comment]
The user isn't clearing the canvas; our code is. `update` calls `clear` and then replots everything.

So, if you do this:

```python
import vcs, cdms2
x = vcs.init()
f = cdms2.open(vcs.sample_data + "/clt.nc")
b = vcs.createboxfill()
s = f("clt")

dp = x.plot(s, b)
x.update()
assert dp.name not in x.display_names
```

you'll see that the display plot is no longer associated with the canvas.

This is an issue for me because I need a specific way of referring to a given plot; Display plots are the most sensible way to do that (they contain all of the relevant information about a plot). The fact that they aren't canonical (and that there is, in fact, no canonical way to keep track of the same plot on a canvas over time), leads to some crazy workarounds.

On top of that, disassociating the objects that we *hand back to the user* from actually having any effect on the world is a pretty bad leaky abstraction.

**Comment by chaosphere2112** [comment]
@aashish24 @doutriaux1 But no, I don't need this in 2.4; it's fine to punt till 3.0, it's just needed for the new GUI.

**Comment by aashish24** [comment]
> On top of that, disassociating the objects that we hand back to the user from actually having any effect on the world is a pretty bad leaky abstraction.

ah I see. Yes, update should not get rid of plots, only clear should. Actually on that note, why we call clear during update. Since we are using VTK, we sould be just update the pipeline. I can take closer look at this post 2.4

**Comment by chaosphere2112** [comment]
> Since we are using VTK, we sould be just update the pipeline. I can take closer look at this post 2.4

:+1:

**Comment by chaosphere2112** [comment]
@aashish24 @doutriaux1 So I need this for the GUI in this release; shall I rebase off of master and make sure it works? Or will someone at kitware do what @aashish24 mentioned earlier and make it so we actually update the pipelines instead of clearing/replotting? Either way, I'm going to need display plots to stick around.

**Comment by chaosphere2112** [comment]
@aashish24 @doutriaux1 Hey, let's merge this soon.

**Comment by aashish24** [comment]
@chaosphere2112 looks like a merge conflict. Should I update the branch?

**Comment by aashish24** [comment]
@chaosphere2112 do we still want this branch?

**Comment by aashish24** [comment]
ping @chaosphere2112

**Comment by chaosphere2112** [comment]
@aashish24 Yup, still want it. Sorry, changed computers and my folder that gets all of my github email was minimized, which naturally means that outlook wouldn't show any notifications that I have anything in it.

**Comment by aashish24** [comment]
> @aashish24 Yup, still want it. Sorry, changed computers and my folder that gets all of my github email was minimized, which naturally means that outlook wouldn't show any notifications that I have anything in it.

no problem, thanks for the update. If you can update the branch with current master we can review it.

**Comment by chaosphere2112** [comment]
@aashish24 Merged master

**Comment by aashish24** [comment]
@chaosphere2112 sorry it has conflict with master now :cry: I missed the window to review this one.

**Comment by chaosphere2112** [comment]
@aashish24 Should be fine to work with now.

**Comment by danlipsa** [comment]
@chaosphere2112 @aashish24 This is a PR that dates from before my time at uvcdat :-), but I have been reading through code related to this for the cdat-web work. So I'll comment on this PR.

File: Packages/vcs/vcs/Canvas.py [hunk 1]
```diff
@@ -2269,7 +2269,7 @@ def drawtextcombined(self, Tt_name=None, To_name=None, string=None,
                        'xbounds', 'ybounds', 'xname', 'yname', 'xunits', 'yunits', 'xweights', 'yweights',
                        'comment1', 'comment2', 'comment3', 'comment4', 'hms', 'long_name', 'zaxis',
                        'zarray', 'zname', 'zunits', 'taxis', 'tarray', 'tname', 'tunits', 'waxis', 'warray',
-                       'wname', 'wunits', 'bg', 'ratio', 'donotstoredisplay', 'render', 'continents_line']
+                       'wname', 'wunits', 'bg', 'ratio', 'donotstoredisplay', 'render', 'continents_line', "display_name"]
```
Comment (danlipsa) [inline Packages/vcs/vcs/Canvas.py]: Can you add a description of what display_name is used for. Not sure where the best place for this is.

File: Packages/vcs/vcs/VTKPlots.py [hunk 2]
```diff
@@ -315,7 +313,7 @@ def configureEvent(self, obj, ev):
             restart_anim = self.canvas.configurator.animation_timer is not None
         else:
             restart_anim = False
-        self.canvas.clear(render=False)
+        self.canvas.clear(render=False, preserve_display=True)
```
Comment (danlipsa) [inline Packages/vcs/vcs/VTKPlots.py]: Can you add a line describing preserve_display in the documentation for clear? Is there a place where all parameters for clear() are listed and described?

**Comment by danlipsa** [comment]
> Yes, update should not get rid of plots, only clear should. Actually on that note, why we call clear during update. Since we are using VTK, we sould be just update the pipeline. I can take closer look at this post 2.4
@aashish24 @chaosphere2112 @doutriaux1 I also would like to look at that code:
https://github.com/UV-CDAT/uvcdat/issues/1868

**Comment by chaosphere2112** [comment]
@danlipsa I added documentation, and tested all of the existing arguments to plot (as well as deprecating a number of ones that aren't that useful and just override variable attributes from the netcdf file).

**Comment by danlipsa** [comment]
@chaosphere2112 Thanks for updating the comments. LGTM if all test pass. We should reexamine this code as discussed earlier.

**Comment by chaosphere2112** [comment]
@danlipsa @doutriaux1 @aashish24 Just ran tests, this was my result:

```
The following tests FAILED:
	115 - test_vcs_boxfill_robinson_wrap (Failed)
	425 - test_vcs_continents (Failed)
	540 - flake8_Thermo (Failed)
	579 - flake8_xmgrace (Failed)
	583 - diags_test_04 (Failed)
	584 - diags_test_41 (Failed)
	585 - diags_test_05 (Failed)
	586 - diags_test_06 (Failed)
	587 - diags_test_07 (Failed)
	588 - diags_test_08 (Failed)
	589 - diags_test_09 (Failed)
	591 - diags_test_11 (Failed)
	593 - diags_test_13 (Failed)
	594 - diags_test_15 (Failed)
```

Are all of those failing in master right now?

**Comment by danlipsa** [comment]
I know about continents and diags which have been failing for some time. I have not seen the other. They were not there a few days ago.

**Comment by chaosphere2112** [comment]
@danlipsa The flake8 ones are probably because I have a "newer" install of flake8 for reasons unbeknownst to me. Wouldn't worry about them. I'll look into the boxfill robinson wrap.

**Comment by aashish24** [comment]
@danlipsa @chaosphere2112 what's the status of this PR?

**Comment by danlipsa** [comment]
@chaosphere2112 What about the test_vcs_boxfill_robinson_wrap

**Comment by chaosphere2112** [comment]
@danlipsa That test just uniformly fails on my machine, so it's unrelated.

**Comment by danlipsa** [comment]
@chaosphere2112 What is the diff between the created image and the baseline? Maybe you should add an additional baseline if the diff is small or file an issue if the diff is large. Anyway, in that case, I am fine with the change.

**Comment by chaosphere2112** [comment]
test_robinson_wrap_diff

The diff is 24; probably can just increase the threshold, but that's a separate pull request.

**Comment by danlipsa** [comment]
I would do a separate baseline rather than increasing the threshold. This way you control what kind of small differences are acceptable.

**Comment by chaosphere2112** [comment]
@danlipsa Regardless, a discussion for a different PR.

**Comment by danlipsa** [comment]
@chaosphere2112 Sounds good. You can go ahead and merge this whenever you are ready.

**Pull request merged**